caikit/runtime/client/remote_model_finder.py

```diff
@@ -0,0 +1,304 @@
+
+        # Type/Value check connection parameters
+        self._connection = ConnectionInfo(**config.connection)
+        self._tls = self._connection.tls
+
+        # Type/Value check default parameters
+        self._model_key = config.get("model_key", MODEL_MESH_MODEL_ID_KEY)
+        error.type_check("<COR72281587E>", str, model_key=self._model_key)
+
+        self._protocol = config.get("protocol", "grpc")
+        error.value_check(
+            "<COR72281545E>",
+            self._protocol in ["grpc", "http"],
+            "Unknown protocol: %s",
+            self._protocol,
+        )
+
+        if self._protocol == "grpc" and self._tls.enabled:
+            error.value_check(
+                "<COR74451567E>",
+                not self._tls.insecure_verify,
+                "GRPC does not support insecure TLS connections."
+                "Please provide a valid CA certificate",
+            )
+
+        # Type/Value check model parameters
+        self._discover_models = config.get("discover_models", True)
+        self._supported_models = config.get("supported_models") or {}
+        self._min_poll_time = config.get("min_poll_time", 30)
+        error.type_check(
+            "<COR74343245E>",
+            dict,
+            supported_models=self._supported_models,
+        )
+        error.type_check(
+            "<COR74342245E>",
+            bool,
+            discover_models=self._discover_models,
+        )
+        error.type_check("", int, min_poll_time=self._min_poll_time)
+
+        # If discovery models is enabled construct lock objects
+        # and then run discovery
+        if self._discover_models:
+            self._last_discovered_time = None
+            self._poll_delta = timedelta(seconds=self._min_poll_time)
+            self._discovery_lock = Lock()
+            self._supported_models.update(self.discover_models())
+
+    def find_model(
+        self,
+        model_path: str,
+        **__,
+    ) -> Optional[RemoteModuleConfig]:
+        """Check if the remote runtime supports the model_path"""
+
+        # If model_path is not detected and discover models is enabled attempt
+        # rediscovery
+        if model_path not in self._supported_models and self._discover_models:
+            self._safe_discover()
+
+        # If model_path is not one of the supported models then raise an error
+        if model_path not in self._supported_models:
+            log.debug(
+                "Model %s is not supported by finder %s",
+                model_path,
+                self._instance_name,
+            )
+            return
+
+        return RemoteModuleConfig.load_from_module(
+            module_reference=self._supported_models.get(model_path),
+            connection_info=self._connection,
+            protocol=self._protocol,
+            model_key=self._model_key,
+            model_path=model_path,
+        )
+
+    def discover_models(self) -> Dict[str, str]:
```

Comment:
If this is intended to always live behind `_safe_discover`, should this have a leading `_` to indicate that it's protected?